opentelemetry-api/src/opentelemetry/trace/__init__.py

```diff
@@ -409,9 +410,80 @@ def use_span(
         yield
 
 
+class ProxyTracer(Tracer):
+    """Proxies all calls to current TracerProvider
+    """
+
+    def __init__(
+        self, get_current_tracer: typing.Callable[[], "Tracer"],
+    ) -> None:
+        self._get_current_tracer = get_current_tracer
+
+    def start_span(
+        self,
+        name: str,
+        context: typing.Optional[Context] = None,
+        kind: SpanKind = SpanKind.INTERNAL,
+        attributes: types.Attributes = None,
+        links: typing.Sequence[Link] = (),
+        start_time: typing.Optional[int] = None,
+        record_exception: bool = True,
+        set_status_on_exception: bool = True,
+    ) -> "Span":
+        return self._get_current_tracer().start_span(
+            name=name,
+            context=context,
+            kind=kind,
+            attributes=attributes,
+            links=links,
+            start_time=start_time,
+            record_exception=record_exception,
+            set_status_on_exception=set_status_on_exception,
+        )
+
+    @contextmanager  # type: ignore
+    def start_as_current_span(
+        self,
+        name: str,
+        context: typing.Optional[Context] = None,
+        kind: SpanKind = SpanKind.INTERNAL,
+        attributes: types.Attributes = None,
+        links: typing.Sequence[Link] = (),
+        start_time: typing.Optional[int] = None,
+        record_exception: bool = True,
+        set_status_on_exception: bool = True,
+    ) -> typing.Iterator["Span"]:
+        with self._get_current_tracer().start_as_current_span(
+            name=name,
+            context=context,
+            kind=kind,
+            attributes=attributes,
+            links=links,
+            start_time=start_time,
+            record_exception=record_exception,
+            set_status_on_exception=set_status_on_exception,
+        ) as span:
+            yield span
+
+    @contextmanager  # type: ignore
+    def use_span(
+        self, span: "Span", end_on_exit: bool = False,
+    ) -> typing.Iterator[None]:
+        with self._get_current_tracer().use_span(
+            span=span, end_on_exit=end_on_exit,
+        ) as context:
+            yield context
+
+
 _TRACER_PROVIDER = None
 
 
+@functools.lru_cache(maxsize=128)  # type: ignore
+def _get_current_tracer(*args: typing.Any, **kwargs: typing.Any) -> "Tracer":
+    tracer_provider = get_tracer_provider()
+    return tracer_provider.get_tracer(*args, **kwargs)  # type: ignore
```

Comment:
I think the user would get a stale `DefaultTracer` if the `cache_clear()` below happened while this body was executing. There were no concurrency guarantees around the globals here (because it was intended to be called once in initialization), so might need some other additions.

Reply:
> I think the user would get a stale DefaultTracer if the cache_clear() below happened while this body was executing. 

And it will definitely will get old tracer if it is called before `set_tracer_provider()`. So yes, all traces that are produced before complete setup of tracing will be lost. Exactly the same way as now.

But the main idea of that PR — all traces after configuration will be handled. Currently there is no way to do that. So I think it's an improvement.


> because it was intended to be called once in initialization

I agree. That should be configured early, during **initialization**, before spawning threads etc. All I want here is to do that initialization after most imports and maybe after few other calls.

Reply:
I mean the `ProxyTracer` will be stuck proxying to a `DefaultTracer` because of the cache. I don't think `cache_clear()` is thread safe. In which case, all spans for that tracer will be lost even after configuration.

1. thread 1 has a proxy tracer call `_get_current_tracer()` for the first time (or when the cache is full). L483 executes right before a context switch
2. thread 2 calls `set_tracer_provider()` which calls `cache_clear()` and empties the caches so proxies will use the real provider
3. `_get_current_tracer()` from step 1. above finishes and overwrites the cache with a `DefaultTracer` for that proxy

> But the main idea of that PR — all traces after configuration will be handled. Currently there is no way to do that. So I think it's an improvement.

Reply:
Hm, valid point.

In python implementation it [uses](https://github.com/python/cpython/blob/7cf0aad96d1d20f07d7f0e374885f327c2d5ff27/Lib/functools.py#L570) lock for bounded cache, but [not for unbounded](https://github.com/python/cpython/blob/7cf0aad96d1d20f07d7f0e374885f327c2d5ff27/Lib/functools.py#L551).

In real world python version overridden [with C](https://github.com/python/cpython/blob/7cf0aad96d1d20f07d7f0e374885f327c2d5ff27/Lib/functools.py#L642) [which](https://github.com/python/cpython/blob/master/Modules/_functoolsmodule.c) does not [Py_BEGIN_ALLOW_THREADS](https://github.com/python/cpython/blob/7cf0aad96d1d20f07d7f0e374885f327c2d5ff27/Modules/_functoolsmodule.c#L877) and uses GIL for synchronization. 

So that may happen if there will be no C-implementation, `set_tracer_provider` will be used before spawning threads, if switch will happen exactly there… I think it's rather safe to do it this way.

Reply:
Nice find. We do support PyPy and I'd rather be safe than sorry regarding when threads are spawned. I believe wsgi/asgi runners could start multiple threads early on outside of the programmer's control?

Reply:
Do we need to use a cache for this? The fact that we need to look at implementation details of different run times makes not want to use it for this purpose. I feel like we are abusing this by using it in a place where we don't really need a cache. 

What if the proxy called `get_tracer_provider()` on every call until it received a non-default one and then used it from there on? 

Something like:


```python
class ProxyTracer:

  def __init__(self, instrumentation_name, instrumentation_version):
    self._real_tracer = None
    self._noop_tracer = NoopTracer()
    self._inst_name = instrumentation_name
    self._inst_version = instrumentation_version

  @property
  def _tracer(self):
    if self._real_tracer:
      return self._real_tracer

    provider = get_tracer_provider()
    if not isinstance(provider, DefaultTracerProvider):
        self._real_tracer = provider.get_tracer(self._inst_name, self._inst_version)
        return self._real_tracer

    return self._noop_tracer

  def start_span(self):
    return self._tracer.start_span()

  # rest of tracer api methods that proxy calls to `self._tracer`
```

Assuming, default/noop tracer provider's `get_tracer()` method would always return a `ProxyTracer` instance. Unless I'm missing something obvious, these two changes should solve this problem no matter whether convenience methods are used or Otel API is directly used.

Reply:
> What if the proxy called get_tracer_provider() on every call until it received a non-default one and then used it from there on?

And what if real won't be provided at all because tracing is disabled for this instance. Traceable instances may work faster than where tracing is disabled?

> Something like:

Didn't you just rewritten cache mechanism by yourself? Then why not to use standard one?

My implementation reuses same tracer instance for all proxies, your creates new for each.

I don't think that is critical to have exactly one instance, but why not if it's so easy

Reply:
> Didn't you just rewritten cache mechanism by yourself? Then why not to use standard one?

No. It's not a (LRU) cache, just a plain wrapper that is very simple and clear to read/understand. When I see something using an LRU cache, it usually means the code has been added as a performance optimization and we can safely assume that the cache can be invalidated at any time. Our problem is not that creating tracers is an expensive operation. Our problem is that we need to be able create a tracer implementation that can lazily upgrade itself from a noop to a real implementation. Just because a solution for another problem "works" doesn't mean we should use it. 

>> What if the proxy called get_tracer_provider() on every call until it received a non-default one and then used it from there on?

> And what if real won't be provided at all because tracing is disabled for this instance. Traceable instances may work faster than where tracing is disabled?

If tracing is disabled somehow then the tracer won't work as expected.